R/format_ISO8601.r

```diff
@@ -0,0 +1,137 @@
+            precision_format <-
+              format_ISO8601_precision_check(precision=precision, max_precision="ymdhms", usetz=usetz)
+            # Note that the usetz argument to as.character is always FALSE because the time zone is handled in the precision argument.
+            sprintf("%s/%s",
+                    as.character(x@start, format=precision_format, usetz=FALSE),
+                    as.character(x@start + x@.Data, format=precision_format, usetz=FALSE))
+          })
+
+#' @rdname format_ISO86901
+#' @export
+setMethod("format_ISO8601", signature="Duration",
+          function(x, usetz=FALSE, precision=NULL, ...) {
+            if (!is.null(precision)) {
+              warning("precision is not used for Duration objects")
+            }
+            sprintf("PT%sS", format(x@.Data))
+          })
+
+#' @rdname format_ISO86901
+#' @export
+setMethod("format_ISO8601", signature="Period",
+          function(x, usetz=FALSE, precision=NULL, ...) {
+            if (!is.null(precision)) {
+              warning("precision is not used for Period objects")
+            }
+            date_part <-
+              paste0(
+                ifelse(x@year != 0,
+                       paste0(x@year, "Y"),
+                       ""),
+                ifelse(x@month != 0,
+                       paste0(x@month, "M"),
+                       ""),
+                ifelse(x@day != 0,
+                       paste0(x@day, "D"),
+                       ""))
+            time_part <-
+              paste0(
+                ifelse(x@hour != 0,
+                       paste0(x@hour, "H"),
+                       ""),
+                ifelse(x@minute != 0,
+                       paste0(x@minute, "M"),
+                       ""),
+                ifelse(x@.Data != 0,
+                       paste0(x@.Data, "S"),
+                       ""))
+            mask_neither <- (nchar(date_part) == 0) & (nchar(time_part) == 0)
+            time_part[mask_neither] <- "0S"
+            ifelse(nchar(time_part),
+                   paste0("P", date_part, "T", time_part),
+                   paste0("P", date_part))
+          })
+
+#' Provide a format for ISO8601 dates and times with the requested precision.
+#'
+#' @param precision The amount of precision to represent with substrings of
+#'   "ymdhms", as "y"ear, "m"onth, "d"ay, "h"our, "m"inute, and "s"econd. (e.g.
+#'   "ymdhm" would show precision through minutes.
+#' @param max_precision The maximum precision allowed to be output.
+#' @param usetz Include the timezone in the output format
+#' @return A format string ready for \code{as.character} methods.
+#' @details
+#' When \code{NULL}, \code{max_precision} is returned.  When \code{precision} is
+#' more precise than \code{max_precision}, a warning is given and
+#' \code{max_precision} is returned.
+format_ISO8601_precision_check <- function(precision, max_precision, usetz=FALSE) {
+  precision_map <- list(y="%Y",
```

Comment:
This should be stored outside of the function to avoid construction on every call.